File: Data Structures & Algorithms/target-sum/submission-2.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {

        int[][] cache = new int[nums.length][target + 2001];
        for (int i = 0; i < nums.length; i++) {
            Arrays.fill(cache[i], -1);
        }

        return dfsHelper(0, nums, target, cache);
    }

    private int dfsHelper(int i, int[] nums, int target, int[][] cache) {
        if (i == nums.length) {
            if (target == 0) {
                return 1;
            }
            return 0;
        }
        if (cache[i][target + 1000] != -1) {
            return cache[i][target + 1000];
        }

        cache[i][target + 1000] = dfsHelper(i + 1, nums, target + nums[i], cache) +
                dfsHelper(i + 1, nums, target - nums[i], cache);

        return cache[i][target + 1000];
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-2.py
class Solution:
    def climbStairs(self, n: int) -> int:
        cache = [0 for i in range(n)]
        return self.dfs(0,n,cache)

    def dfs(self, i: int,n: int,cache: List[int]) -> int:
        if(i==n):
            return 1
        if(i>n):
            return 0
        if(cache[i]!=0):
            return cache[i]
        
        cache[i] = self.dfs(i+1,n,cache) + self.dfs(i+2,n,cache)

        return cache[i]